lib/Sema/ConstraintSystem.h

```diff
@@ -3516,22 +3543,51 @@ class DisjunctionChoiceProducer : public BindingProducer<DisjunctionChoice> {
         IsExplicitConversion(disjunction->isExplicitConversion()) {
     assert(disjunction->getKind() == ConstraintKind::Disjunction);
     assert(!disjunction->shouldRememberChoice() || disjunction->getLocator());
+
+    // Order and partition the disjunction choices.
+    partitionDisjunction();
   }
 
   DisjunctionChoiceProducer(ConstraintSystem &cs,
                             ArrayRef<Constraint *> choices,
                             ConstraintLocator *locator, bool explicitConversion)
       : BindingProducer(cs, locator), Choices(choices),
-        IsExplicitConversion(explicitConversion) {}
+        IsExplicitConversion(explicitConversion) {
+
+    // Order and partition the disjunction choices.
+    partitionDisjunction();
+  }
 
   Optional<Element> operator()() override {
     unsigned currIndex = Index;
     if (currIndex >= Choices.size())
       return None;
 
     ++Index;
-    return DisjunctionChoice(currIndex, Choices[currIndex],
-                             IsExplicitConversion);
+
+    bool endOfPartition = false;
+
+    if (PartitionEnd[PartitionIndex] == Index) {
```

Comment:
Maybe instead of `isEndOfDisjunctionPartition` it would be better to rename `needsToComputeNext()` into `endOfPartitition()` and use it for both producers? That way `BindingStep` could just implement non-virtual `shouldStopAfter` as `AnySolved && Producer.endOfPartition()` and this `if` block could be moved there, WDYT?

Reply:
Sure. I'll push a change for that.